psi4/driver/driver.py

```diff
@@ -1932,35 +1932,165 @@ def gdma(wfn, datafile=""):
         os.remove(commands)
 
 
-def fchk(wfn, filename):
+def fchk(wfn, filename, debug=False):
     """Function to write wavefunction information in *wfn* to *filename* in
     Gaussian FCHK format.
 
     .. versionadded:: 0.6
 
     :returns: None
 
+    :type wfn: :py:class:`~psi4.core.Wavefunction`
+    :param wfn: set of molecule, basis, orbitals from which to generate fchk file
+
     :type filename: string
     :param filename: destination file name for FCHK file
 
-    :type wfn: :py:class:`~psi4.core.Wavefunction`
-    :param wfn: set of molecule, basis, orbitals from which to generate fchk file
+    :type debug: boolean
+    :param debug: returns a dictionary to aid with debugging
 
     Notes
     -----
     * A description of the FCHK format is http://wild.life.nctu.edu.tw/~jsyu/compchem/g09/g09ur/f_formchk.htm
     * The allowed headers for methods are general and limited, i.e., "Total SCF|MP2|CI|CC Density",
-      so "CC" is always used for the post-HF case.
+      PSI4 will try to find the right one for the current calculation.
+    * Not all theory modules in PSI4 are compatible with the FCHK writer. Currently available and tested are:
+      All SCF/DFT variants (if gradient avaiable), DF-RHF-MP2 (dfmp2 module), CONV-RHF/UHF-CC methods (ccenergy module), orbital-relaxed DCT (ODC-* functionals) (dct module). No UHF-MP2 densities are available.
+      A warning will be printed if a theory module is not supported.
 
     :examples:
 
     >>> # [1] FCHK file for DFT calculation
     >>> E, wfn = energy('b3lyp', return_wfn=True)
     >>> fchk(wfn, 'mycalc.fchk')
 
+    >>> # [2] FCHK file for correlated densities
+    >>> E, wfn = gradient('ccsd', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk')
+
     """
-    fw = core.FCHKWriter(wfn)
+    #NOTE: This function should be overhauled once all modules set Ca/Cb, Da/Db in the same manner.
+
+    # QC_MODULE overview for FCHK
+    # OCC: (includes both occ/dfocc modules
+    #  wfn.D() = SCF
+    #  wfn.C() = SCF
+    #  refwfn.D() = SCF
+    # -> not allowed, graceful exit
+    # DCT:
+    #  wfn.D() = SCF
+    #  wnf.C() = DCT
+    #  refwfn.D() = SCF
+    # -> allowed, but FCHK entries are switched around
+    # CCENERGY
+    #  wfn.D() = CC
+    #  wfn.C() = SCF 
+    #  refwfn.D() = CC
+    # -> allowed, will recompute SCF density from wfn.C()
+    # DFMP2:
+    #  wfn.D() = MP2
+    #  wnf.C() = NO(MP2)
+    #  refwfn.D() = SCF 
+    # -> allowed, will take refwfn.D() for SCF density
+
+    allowed = ['DFMP2','SCF','CCENERGY','DCT']
+    module_ = wfn.module().upper()
+    if (module_ not in allowed):
+        core.print_out(f"Theory module {module_} is currently not supported by the FCHK writer.")
+        return None
+
+    # handle special cases
+    if (module_ == 'CCENERGY'):
+        fw = core.FCHKWriter(wfn)
+        fw.recompute_scf_density(True)
+    elif (module_ == 'DCT'):
```

Comment:
Looking at the code, I don't understand how DCT FCHK works. Just to confirm, have you tested DCT FCHK? I don't see that anyway in your tests. The reasons for my skepticism are:
1. You're re-computing the SCF density and then writing that back as the density matrix of `wfn_`, not its reference wavefunction. How does the correlated density matrix get passed to the FCHKWriter?
2. The orbitals you're using to compute the reference density come from the input wavefunction's `Ca` and `Cb`, not the reference wavefunction's. Are those the same? I would have thought that the input wavefunction's orbitals would have been the optimized orbitals computed during the iterations.

Reply:
I tested by comparing against a ccsd fchk file, though CC and DCT density are different, they are both different to the SCF density. I can add a regression test. Not sure if I can get a DCT density matrix out of the `dct` module?

1. `wfn_.Da().np[:] = Da` writes the new density to `wfn_` and from there FCHKWriter takes it.
2. I calculate the correlated density from the input wfn Ca/Cbs, which are the optimized DCT orbitals.

The writer needs `wfn.D()` be the DCT density, but it's the SCF density on input. I dont think orbital-optimized methods are covered in FCHK format and the orbital coefficient field needs come from the SCF. So there is this confusing switching around happening.

Reply:
Okay, I see.

That formula for going from the orbitals to the density is correct for mean-field theories (HF and DFT), but **wrong** for anything else. What you need is the DCT one-particle density matrix (OPDM) available on the wavefunction. I can get a PR fixing that (for any ODC-06, ODC-12, and ODC-13, and also DC-06 with a gradient computation) tomorrow. I'd need it anyways for Gradient Refactor. Getting OPDM for any other DCT methods would require deriving analytic gradient expressions, and that isn't worthwhile.

Reply:
right! Clearly not paying enough attention to all this :( 
If you find no time for a PR I can also disable DCT for now.

Reply:
Unless a miracle happens, this is actually the perfect time for me to work on a PR like this.